### PR TITLE
Annotator update

### PR DESCRIPTION
**Pull request by maciejdrozdowski**
1) Paper.js Groups & Compound Path Major Changes:
- Add CategoryGroup class, which represent object with multiple Annotations
- Add AnnotationGroup class, which hold references for Annotation Shape & Annotation KeypointsGroup
- Add AnnotationShape class, which represent Annotation CompoundPath (with possibility to mark it as BBOX)
- Add KeypointsGroup class, which hold Annotation Keypoints and edges ( lines ) between them
- Add Keypoint class, which represent single Keypoint circle

2) BBox Feature:

(a) Adjust AnnotationShape to support BBOX property

(b) AnnotationShape with BBOX property represent rectangle shape and behave differently than normal shape

(c) useSelect Tool
- In case click + drag (on BBOX POINT!)-> we resize WHOLE rectangle rather than move point
- In case click + drag (on BBOX FILL!) -> we move WHOLE rectangle

(d) useBBOX Tool
- add new shape -> old shape is removed and replaced with AnnotationShape with BBOX property
- add new shape (in case AnnotationShape BBOX = true) -> prevent from doing anything

(e) usePolygon/useBrush Tool
- unite/subtract new shape into AnnotationShape with BBOX property -> AnnotationShape lost ALL BBOX behaviors and behave as regular path

3) Keypoint Feature:
(a) useKeypoint Tool can add new Keypoints on Project

(b) useSelect Tool
- on click - select Keypoint in some KeypointGroup
- on click on select Keypoint -> unselect Keypoint
- on click on different Keypoint (when some other Keypoint is selected) -> connect two Keypoints with line  / or remove line if already existed
- on click + drag -> move Keypoint to new position & update existing lines between other Keypoints
- on click on selected Keypoint + shift-> remove Keypoint & all existing lines to it

(c) Change format of Keypoint export type

4) Major cleaning for Tools hooks:
- Move AnnotationGroup plain functions into groupUtils
- Move AnnotationShape plain functions into shapeUtils
- Move SelectTool plain functions into selectUtils

5) Major cleaning for useSelect hook - refactor all vague names and group them based on SELECT / HOVER functionality

6) Minor cleaning for useGroup hook - separate useShape and useKeypoint sub-Hooks for managing specific things

7) Manage Annotation Feature:
- Move CategoryInfo and AnnotationInfo to Annotator.types
- Adjust useInfo hook to create / remove annotations ( split it into subHooks )
- Adjust useFilter hook to return Dictionary instead of array
- Adjust useChoices hook setSelected hook
- Adjust useDataset hook response
- Add some Config initial values

### CODE REVIEW

File: client/src/Annotator/Annotator.tsx [hunk 1]
```diff
@@ -27,10 +28,9 @@ const Annotator: React.FC<{ imageId: number }> = ({ imageId }) => {
         selected,
         setSelected,
     } = useChoices();
-
-    const { categories, filename, previous, next } = useDataset(imageId);
-    const info = useInfo(categories);
-    const filter = useFilter(categories);
+    const dataset = useDataset(imageId);
```
Comment (wojciech-bilicki) [inline client/src/Annotator/Annotator.tsx]: I believe that you can unpack dataset here, so you don't repeat yourself later on

File: client/src/Annotator/Info/useDataset.tsx [hunk 2]
```diff
@@ -10,23 +10,36 @@ import * as AnnotatorApi from '../annotator.api';
...
     const [, dispatch] = useGlobalContext();
 
     // datasetdata
     const [dataset, setDataset] = useState<Maybe<Dataset>>(null);
     const [categories, setCategories] = useState<Array<Category>>([]);
+    const [isLoading, _setIsLoading] = useState<boolean>(false);
 
     // image data
-    const [metadata, setMetadata] = useState<Maybe<{}>>(null);
+    const [, setMetadata] = useState<Maybe<{}>>(null);
```
Comment (wojciech-bilicki) [inline client/src/Annotator/Info/useDataset.tsx]: Any chance we can tell the type here more precisely than just an empty object?

File: client/src/Annotator/Info/useDataset.tsx [hunk 3]
```diff
@@ -10,23 +10,36 @@ import * as AnnotatorApi from '../annotator.api';
...
     const [filename, setFilename] = useState<string>('');
     const [next, setNext] = useState<Maybe<number>>(null);
     const [previous, setPrevious] = useState<Maybe<number>>(null);
-    const [categoriesIds, setCategoriesIds] = useState<Array<number>>([]);
-    const [annotating, setAnnotating] = useState<[]>([]);
+    const [, setCategoriesIds] = useState<Array<number>>([]);
+    const [, setAnnotating] = useState<[]>([]);
```
Comment (wojciech-bilicki) [inline client/src/Annotator/Info/useDataset.tsx]: ditto

File: client/src/Annotator/Paper/Utils/selectUtils.tsx [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    return { text, box };
+};
+
+const createShapeText = (
+    categoryId: Maybe<number>,
+    annotationId: Maybe<number>,
+) => {
+    // info text
+    let msg;
```
Comment (wojciech-bilicki) [inline client/src/Annotator/Paper/Utils/selectUtils.tsx]: you don't modify `msg` after the initial lines why not declare it as const and initialise it with proper value form the very beginning?

**Review by wojciech-bilicki — CHANGES_REQUESTED** [review]

**Review by wojciech-bilicki — APPROVED** [review]

**Review by wiola-dreszler — APPROVED** [review]

**Pull request merged**